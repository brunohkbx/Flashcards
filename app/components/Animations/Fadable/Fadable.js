import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Animated } from 'react-native';

class Fadable extends Component {
  state = {
    visible: true,
    opacity: new Animated.Value(1)
  }

  onComplete = () => this.setState(
    { visible: false }, this.props.onAnimationEnd
  );

  componentDidUpdate(prevProps) {
    const { fade } = this.props;

    if (fade && prevProps.fade !== fade) {
      const { opacity } = this.state;

      Animated.timing(
        opacity,
        { toValue: 0, duration: 350, useNativeDriver: true }
      ).start(this.onComplete);
    }
  }

  render() {
    const { opacity, visible } = this.state;
    const { children } = this.props;

    return (
      <Animated.View
        style={{
          opacity: opacity.interpolate({
            inputRange: [0, 1], outputRange: [0, 1]
          })
        }}
      >
        {visible && children}
      </Animated.View>
    );
  }
}

Fadable.propTypes = {
  fade: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onAnimationEnd: PropTypes.func
};

Fadable.defaultProps = {
  fade: false
};

export default Fadable;
