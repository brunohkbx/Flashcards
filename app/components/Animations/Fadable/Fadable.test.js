import React from 'react';
import { shallow } from 'enzyme';
import Fadable from './Fadable';
import { Animated, View } from 'react-native';

describe('Fadable', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      fade: false,
      onAnimationEnd: () => jest.fn()
    }, propOverrides);

    const wrapper = shallow(<Fadable {...defaultProps} ><View /></Fadable>);
    const wrapperInstance = wrapper.instance();

    return {
      wrapper, wrapperInstance
    };
  };

  beforeEach(() => {
    jest.restoreAllMocks()
  });

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe('onComplete', () => {
    it('sets visible to false and passes onAnimationEnd as callback to setState', () => {
      const { wrapperInstance } = setup();
      jest.spyOn(wrapperInstance, 'setState');

      wrapperInstance.onComplete();

      expect(wrapperInstance.setState).toHaveBeenCalledWith(
        { visible: false }, wrapperInstance.props.onAnimationEnd
      );
    });
  });

  describe('componentDidUpdate', async () => {
    describe('When component is faded', () => {
      it('starts the animation', () => {
        const { wrapper, wrapperInstance } = setup();
        jest.spyOn(Animated, 'timing');

        wrapper.setProps({ fade: true });

        expect(Animated.timing).toHaveBeenCalledWith(
          wrapperInstance.state.opacity,
          { toValue: 0, duration: 350, useNativeDriver: true  }
        );
      });

      it('calls onComplete when animation ends', () => {
        const { wrapper, wrapperInstance } = setup();
        const mockAnimatedStart = jest.fn();
        jest.spyOn(Animated, 'timing')
          .mockImplementation(() => ({ start: mockAnimatedStart }));

        wrapper.setProps({ fade: true });

        expect(
          mockAnimatedStart
        ).toHaveBeenCalledWith(wrapperInstance.onComplete);
      });
    });

    describe('When component is not faded', () => {
      it('does not start the animation', () => {
        const { wrapper } = setup();
        jest.spyOn(Animated, 'timing');

        wrapper.setProps({ fade: false });

        expect(Animated.timing).not.toHaveBeenCalled();
      });
    });
  });
});
