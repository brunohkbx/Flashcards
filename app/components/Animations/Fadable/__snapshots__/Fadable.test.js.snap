// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fadable renders properly 1`] = `
<AnimatedComponent
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Component />
</AnimatedComponent>
`;
